File: functions/own_profile/update_my_profile.py
from firebase_admin import firestore
from flask import abort
from models.constants import (
    Collections,
    FriendshipFields,
    GroupFields,
    InvitationFields,
    ProfileFields,
)
from models.data_models import Insights, ProfileResponse
from utils.logging_utils import get_logger


def update_profile(request):
    """
    Updates the authenticated user's profile information.

    This function:
    1. Checks if a profile exists for the authenticated user
    2. Updates the profile with the provided data
    3. If username, name, or avatar changes, updates these fields in related collections:
       - Invitations
       - Friendships (both as sender and receiver)
       - Groups (in member_profiles)

    Args:
        request: The Flask request object containing:
                - user_id: The authenticated user's ID (attached by authentication middleware)
                - validated_params: Profile data that can include:
                    - username: Optional updated username
                    - name: Optional updated display name
                    - avatar: Optional updated avatar URL
                    - location: Optional updated location information
                    - birthday: Optional updated birthday in ISO format
                    - notification_settings: Optional updated list of notification preferences

    Returns:
        A ProfileResponse containing the updated profile information

    Raises:
        404: Profile not found
    """
    logger = get_logger(__name__)
    logger.info(f"Starting update_profile operation for user ID: {request.user_id}")

    # Get the authenticated user ID from the request
    current_user_id = request.user_id

    # Get the validated profile data
    profile_data_input = request.validated_params

    # Initialize Firestore client
    db = firestore.client()

    # Check if profile exists
    profile_ref = db.collection(Collections.PROFILES).document(current_user_id)
    profile_doc = profile_ref.get()

    if not profile_doc.exists:
        logger.warning(f"Profile not found for user {current_user_id}")
        abort(404, description=f"Profile not found")

    # Get current profile data
    current_profile_data = profile_doc.to_dict()
    logger.info(f"Retrieved current profile data for user {current_user_id}")

    # Check if username, name, or avatar has changed
    username_changed = (
        hasattr(profile_data_input, "username")
        and profile_data_input.username is not None
        and profile_data_input.username
        != current_profile_data.get(ProfileFields.USERNAME)
    )

    name_changed = (
        hasattr(profile_data_input, "name")
        and profile_data_input.name is not None
        and profile_data_input.name != current_profile_data.get(ProfileFields.NAME)
    )

    avatar_changed = (
        hasattr(profile_data_input, "avatar")
        and profile_data_input.avatar is not None
        and profile_data_input.avatar != current_profile_data.get(ProfileFields.AVATAR)
    )

    # Prepare update data
    profile_updates = {}

    # Only update fields that are provided in the request
    if (
        hasattr(profile_data_input, "username")
        and profile_data_input.username is not None
    ):
        profile_updates[ProfileFields.USERNAME] = profile_data_input.username

    if hasattr(profile_data_input, "name") and profile_data_input.name is not None:
        profile_updates[ProfileFields.NAME] = profile_data_input.name

    if hasattr(profile_data_input, "avatar") and profile_data_input.avatar is not None:
        profile_updates[ProfileFields.AVATAR] = profile_data_input.avatar

    if (
        hasattr(profile_data_input, "location")
        and profile_data_input.location is not None
    ):
        profile_updates[ProfileFields.LOCATION] = profile_data_input.location

    if (
        hasattr(profile_data_input, "birthday")
        and profile_data_input.birthday is not None
    ):
        profile_updates[ProfileFields.BIRTHDAY] = profile_data_input.birthday

    if (
        hasattr(profile_data_input, "notification_settings")
        and profile_data_input.notification_settings is not None
    ):
        profile_updates[ProfileFields.NOTIFICATION_SETTINGS] = (
            profile_data_input.notification_settings
        )

    # Create a batch for all updates (profile and references)
    batch = db.batch()

    # Update the profile in the batch
    if profile_updates:
        batch.update(profile_ref, profile_updates)
        logger.info(f"Added profile update to batch for user {current_user_id}")

    # If username, name, or avatar changed, update references in other collections
    if username_changed or name_changed or avatar_changed:
        logger.info(
            f"Updating username/name/avatar references for user {current_user_id}"
        )

        # 1. Update all invitations created by this user
        invitations_query = db.collection(Collections.INVITATIONS).where(
            InvitationFields.SENDER_ID, "==", current_user_id
        )

        invitation_docs = invitations_query.stream()
        for invitation_doc in invitation_docs:
            invitation_ref = invitation_doc.reference
            invitation_updates = {}

            if username_changed:
                invitation_updates[InvitationFields.USERNAME] = profile_updates[
                    ProfileFields.USERNAME
                ]

            if name_changed:
                invitation_updates[InvitationFields.NAME] = profile_updates[
                    ProfileFields.NAME
                ]

            if avatar_changed:
                invitation_updates[InvitationFields.AVATAR] = profile_updates[
                    ProfileFields.AVATAR
                ]

            if invitation_updates:
                batch.update(invitation_ref, invitation_updates)

        # 2. Update friendships where user is sender
        friendships_as_sender_query = db.collection(Collections.FRIENDSHIPS).where(
            FriendshipFields.SENDER_ID, "==", current_user_id
        )

        friendship_sender_docs = friendships_as_sender_query.stream()
        for friendship_doc in friendship_sender_docs:
            friendship_ref = friendship_doc.reference
            friendship_updates = {}

            if username_changed:
                friendship_updates[FriendshipFields.SENDER_USERNAME] = profile_updates[
                    ProfileFields.USERNAME
                ]

            if name_changed:
                friendship_updates[FriendshipFields.SENDER_NAME] = profile_updates[
                    ProfileFields.NAME
                ]

            if avatar_changed:
                friendship_updates[FriendshipFields.SENDER_AVATAR] = profile_updates[
                    ProfileFields.AVATAR
                ]

            if friendship_updates:
                batch.update(friendship_ref, friendship_updates)

        # 3. Update friendships where user is receiver
        friendships_as_receiver_query = db.collection(Collections.FRIENDSHIPS).where(
            FriendshipFields.RECEIVER_ID, "==", current_user_id
        )

        friendship_receiver_docs = friendships_as_receiver_query.stream()
        for friendship_doc in friendship_receiver_docs:
            friendship_ref = friendship_doc.reference
            friendship_updates = {}

            if username_changed:
                friendship_updates[FriendshipFields.RECEIVER_USERNAME] = (
                    profile_updates[ProfileFields.USERNAME]
                )

            if name_changed:
                friendship_updates[FriendshipFields.RECEIVER_NAME] = profile_updates[
                    ProfileFields.NAME
                ]

            if avatar_changed:
                friendship_updates[FriendshipFields.RECEIVER_AVATAR] = profile_updates[
                    ProfileFields.AVATAR
                ]

            if friendship_updates:
                batch.update(friendship_ref, friendship_updates)

        # 4. Update groups where user is a member
        # First, get all groups the user is a member of
        groups_query = db.collection(Collections.GROUPS).where(
            GroupFields.MEMBERS, "array_contains", current_user_id
        )

        group_docs = groups_query.stream()
        for group_doc in group_docs:
            group_ref = group_doc.reference
            group_data = group_doc.to_dict()

            # Find and update the user's profile in member_profiles array
            member_profiles = group_data.get(GroupFields.MEMBER_PROFILES, [])
            for i, member_profile in enumerate(member_profiles):
                if member_profile.get(ProfileFields.USER_ID) == current_user_id:
                    # Create update path for the specific array element
                    if username_changed:
                        batch.update(
                            group_ref,
                            {
                                f"{GroupFields.MEMBER_PROFILES}.{i}.{ProfileFields.USERNAME}": profile_updates[
                                    ProfileFields.USERNAME
                                ]
                            },
                        )

                    if name_changed:
                        batch.update(
                            group_ref,
                            {
                                f"{GroupFields.MEMBER_PROFILES}.{i}.{ProfileFields.NAME}": profile_updates[
                                    ProfileFields.NAME
                                ]
                            },
                        )

                    if avatar_changed:
                        batch.update(
                            group_ref,
                            {
                                f"{GroupFields.MEMBER_PROFILES}.{i}.{ProfileFields.AVATAR}": profile_updates[
                                    ProfileFields.AVATAR
                                ]
                            },
                        )

                    break

    # Commit all the updates in a single atomic operation
    if profile_updates or username_changed or name_changed or avatar_changed:
        batch.commit()
        logger.info(f"Committed batch updates for user {current_user_id}")

    # Get the updated profile data
    updated_profile_doc = profile_ref.get()
    updated_profile_data = updated_profile_doc.to_dict()

    # Get insights data
    insights_doc = next(
        profile_ref.collection(Collections.INSIGHTS).limit(1).stream(), None
    )
    insights_data = insights_doc.to_dict() if insights_doc else {}

    # Construct and return the profile response
    return ProfileResponse(
        user_id=current_user_id,
        username=updated_profile_data.get(ProfileFields.USERNAME, ""),
        name=updated_profile_data.get(ProfileFields.NAME, None),
        avatar=updated_profile_data.get(ProfileFields.AVATAR, None),
        location=updated_profile_data.get(ProfileFields.LOCATION, None),
        birthday=updated_profile_data.get(ProfileFields.BIRTHDAY, None),
        notification_settings=updated_profile_data.get(
            ProfileFields.NOTIFICATION_SETTINGS, None
        ),
        summary=updated_profile_data.get(ProfileFields.SUMMARY, None),
        suggestions=updated_profile_data.get(ProfileFields.SUGGESTIONS, None),
        insights=Insights(
            emotional_overview=insights_data.get("emotional_overview", ""),
            key_moments=insights_data.get("key_moments", ""),
            recurring_themes=insights_data.get("recurring_themes", ""),
            progress_and_growth=insights_data.get("progress_and_growth", ""),
        ),
    )

File: functions/src/own_profile/update-my-profile.ts
import { Request } from 'express';
import { getFirestore, PartialWithFieldValue, Timestamp } from 'firebase-admin/firestore';
import { ApiResponse, EventName, ProfileEventParams } from '../models/analytics-events.js';
import { Collections } from '../models/constants.js';
import { ProfileResponse, UpdateProfilePayload } from '../models/data-models.js';
import { pf, ProfileDoc } from '../models/firestore/index.js';
import { NotificationSetting, NudgingSettings, Personality, Tone } from '../models/firestore/profile-doc.js';
import { ConflictError } from '../utils/errors.js';
import { getLogger } from '../utils/logging-utils.js';
import {
  extractConnectToForAnalytics,
  extractGoalForAnalytics,
  extractNudgingOccurrence,
  formatProfileResponse,
  getProfileDoc,
  getProfileInsights,
} from '../utils/profile-utils.js';
import { updateTimeBucketMembership } from '../utils/timezone-utils.js';

import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Updates the authenticated user's profile information.
 *
 * This function:
 * 1. Checks if a profile exists for the authenticated user
 * 2. Updates the profile with the provided data
 * 3. Updates time bucket membership if nudging settings change
 *
 * Note: Denormalized data updates are handled by the profile update trigger.
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 *              - validated_params: Profile data that can include:
 *                - username: Optional updated username
 *                - name: Optional updated display name
 *                - avatar: Optional updated avatar URL
 *                - location: Optional updated location information
 *                - birthday: Optional updated birthday in ISO format
 *                - notification_settings: Optional updated list of notification preferences
 *                - nudging_settings: Optional updated list of nudging preferences
 *                - gender: Optional updated gender information
 *                - goal: Optional updated goal
 *                - connect_to: Optional updated connect_to preferences
 *                - personality: Optional updated personality type
 *                - tone: Optional updated tone preference
 *                - phone_number: Optional updated phone number
 *
 * @returns A ProfileResponse containing the updated profile information
 *
 * @throws 404: Profile not found
 */
export const updateProfile = async (req: Request): Promise<ApiResponse<ProfileResponse>> => {
  const currentUserId = req.userId;
  logger.info(`Starting update_profile operation for user ID: ${currentUserId}`);

  const db = getFirestore();
  const profileData = req.validated_params as UpdateProfilePayload;

  // Get the profile document using the utility function
  const { ref: profileRef, data: currentProfileData } = await getProfileDoc(currentUserId);
  logger.info(`Retrieved current profile data for user ${currentUserId}`);

  // Check if nudging settings have changed
  const nudgingSettingsChanged = profileData.nudging_settings !== undefined;
  const currentTimezone = currentProfileData.timezone;

  // Check if phone number changed
  const phoneChanged =
    profileData.phone_number !== undefined && profileData.phone_number !== currentProfileData.phone_number;

  // If phone number is changing, ensure new phone is not taken
  if (phoneChanged) {
    const existingPhoneDoc = await db
      .collection(Collections.PHONES)
      .doc(profileData.phone_number as string)
      .get();
    if (existingPhoneDoc.exists && existingPhoneDoc.data()?.user_id !== currentUserId) {
      throw new ConflictError('Phone number is already registered by another user');
    }
  }

  // Prepare update data
  const profileUpdates: PartialWithFieldValue<ProfileDoc> = {};

  // Only update fields that are provided in the request
  if (profileData.username !== undefined) {
    profileUpdates[pf('username')] = profileData.username;
  }
  if (profileData.name !== undefined) {
    profileUpdates[pf('name')] = profileData.name;
  }
  if (profileData.avatar !== undefined) {
    profileUpdates[pf('avatar')] = profileData.avatar;
  }
  if (profileData.birthday !== undefined) {
    profileUpdates[pf('birthday')] = profileData.birthday;
  }
  if (profileData.notification_settings !== undefined) {
    profileUpdates[pf('notification_settings')] = profileData.notification_settings as NotificationSetting[];
  }
  if (profileData.nudging_settings !== undefined) {
    profileUpdates[pf('nudging_settings')] = profileData.nudging_settings as NudgingSettings;
  }
  if (profileData.gender !== undefined) {
    profileUpdates[pf('gender')] = profileData.gender;
  }
  if (profileData.goal !== undefined) {
    profileUpdates[pf('goal')] = profileData.goal;
  }
  if (profileData.connect_to !== undefined) {
    profileUpdates[pf('connect_to')] = profileData.connect_to;
  }
  if (profileData.personality !== undefined) {
    profileUpdates[pf('personality')] = profileData.personality as Personality;
  }
  if (profileData.tone !== undefined) {
    profileUpdates[pf('tone')] = profileData.tone as Tone;
  }
  if (profileData.phone_number !== undefined) {
    profileUpdates[pf('phone_number')] = profileData.phone_number;
  }

  // Create a batch for all updates
  const batch = db.batch();

  // Update the profile in the batch
  if (Object.keys(profileUpdates).length > 0) {
    profileUpdates[pf('updated_at')] = Timestamp.now();
    batch.update(profileRef, profileUpdates);
    logger.info(`Added profile update to batch for user ${currentUserId}`);
  }

  // Handle time bucket membership if nudging settings changed and user has timezone
  if (nudgingSettingsChanged && currentTimezone) {
    const newNudgingSettings = profileData.nudging_settings as NudgingSettings;
    await updateTimeBucketMembership(currentUserId, newNudgingSettings, currentTimezone, batch, db);
  }

  // Commit all the updates in a single atomic operation
  if (Object.keys(profileUpdates).length > 0 || nudgingSettingsChanged) {
    await batch.commit();
    logger.info(`Committed batch updates for user ${currentUserId}`);
  }

  // Get the updated profile data
  const updatedProfileDoc = await profileRef.get();
  const updatedProfileData = updatedProfileDoc.data();
  if (!updatedProfileData) {
    throw new Error(`Profile data is undefined for user ${currentUserId}`);
  }

  // Get insights data
  const insightsData = await getProfileInsights(profileRef);

  // Format and return the response
  const response = formatProfileResponse(currentUserId, updatedProfileData, insightsData);

  // Track profile update event
  const event: ProfileEventParams = {
    has_name: !!updatedProfileData.name,
    has_avatar: !!updatedProfileData.avatar,
    has_location: !!updatedProfileData.location,
    has_birthday: !!updatedProfileData.birthday,
    has_notification_settings:
      Array.isArray(updatedProfileData.notification_settings) && updatedProfileData.notification_settings.length > 0,
    nudging_occurrence: extractNudgingOccurrence(updatedProfileData),
    has_gender: !!updatedProfileData.gender,
    goal: extractGoalForAnalytics(updatedProfileData),
    connect_to: extractConnectToForAnalytics(updatedProfileData),
    personality: updatedProfileData.personality || '',
    tone: updatedProfileData.tone || '',
  };

  return {
    data: response,
    status: 200,
    analytics: {
      event: EventName.PROFILE_UPDATED,
      userId: currentUserId,
      params: event,
    },
  };
};
